pkg/stacks/unpack.go

```diff
@@ -17,320 +17,439 @@ limitations under the License.
+func (sp *StackPackage) SetRBAC(rbac v1alpha1.PermissionsSpec) {
+	sp.Stack.Spec.Permissions = rbac
+}
 
-	// read the icon file and encode to base64
-	icons, err := readIcons(fs, root)
-	if err != nil {
-		return "", err
-	}
-	stackRecord.Spec.AppMetadataSpec.Icons = icons
+// GotApp reveals if the AppMetadataSpec has been set
+func (sp *StackPackage) GotApp() bool {
+	return sp.appSet
+}
 
-	// read the RBAC information
-	permissionsObj := v1alpha1.PermissionsSpec{}
-	if err := readFileIntoObject(fs, root, permissionsFileName, false /* not required */, &permissionsObj); err != nil {
-		return "", err
-	}
-	stackRecord.Spec.Permissions = permissionsObj
+// AddGroup adds a group to the StackPackage
+func (sp *StackPackage) AddGroup(path string, sg StackGroup) {
+	sp.Groups[path] = sg
+}
 
-	// marshal the full stack record to yaml and write it to the output
-	stackRecordRaw, err := yaml.Marshal(stackRecord)
-	if err != nil {
-		return "", err
-	}
-	_, err = output.WriteString(yamlSeparator + string(stackRecordRaw))
-	if err != nil {
-		return "", err
-	}
+// AddResource adds a resource to the StackPackage
+func (sp *StackPackage) AddResource(filepath string, sr StackResource) {
+	sp.Resources[filepath] = sr
+}
 
-	return output.String(), nil
+// AddUI adds a resource to the StackPackage
+func (sp *StackPackage) AddUI(filepath string, ui string) {
+	sp.UISpecs[filepath] = ui
 }
 
-func readFileIntoObject(fs afero.Fs, root, fileName string, required bool, obj interface{}) error {
-	file, err := fs.Open(filepath.Join(root, fileName))
-	if err != nil {
-		if os.IsNotExist(err) && !required {
-			// the given file doesn't exist, but it's also not required, this is OK
-			return nil
+// AddIcon adds an icon to the StackPackage
+func (sp *StackPackage) AddIcon(path string, icon v1alpha1.IconSpec) {
+	// TODO(displague) do we want to keep all icons in the Stack spec or just the preferred type?
+	sp.Stack.Spec.AppMetadataSpec.Icons = append(sp.Stack.Spec.AppMetadataSpec.Icons, icon)
+
+	// highest priority wins per path
+	iconMimePriority := map[string]int{"image/svg+xml": 4, "image/png": 3, "image/jpeg": 2, "image/gif": 1}
+	if existing, found := sp.Icons[filepath.Dir(path)]; found {
+		if iconMimePriority[existing.MediaType] > iconMimePriority[icon.MediaType] {
+			return
 		}
-		return err
 	}
-	defer func() { _ = file.Close() }()
 
-	b, err := ioutil.ReadAll(file)
-	if err != nil {
-		return err
+	sp.Icons[filepath.Dir(path)] = &icon
+}
+
+// AddCRD appends a CRD to the CRDs of this StackPackage
+// The CRD will be annotated before being added and the Stack will track ownership of this CRD.
+func (sp *StackPackage) AddCRD(path string, crd *apiextensions.CustomResourceDefinition) {
+	if crd.ObjectMeta.Annotations == nil {
+		crd.ObjectMeta.Annotations = map[string]string{}
+	}
+	crd.ObjectMeta.Annotations[annotationKubernetesManagedBy] = "stack-manager"
+
+	crdGVK := schema.GroupVersionKind{
+		Group:   crd.Spec.Group,
+		Version: crd.Spec.Version,
```

Comment:
Version has been deprecated: https://github.com/kubernetes/apiextensions-apiserver/blob/master/pkg/apis/apiextensions/v1beta1/types.go#L44-L49

Some CRDs may have multiple versions.  We'd either need to track all of them, or maybe just track the "[stored](https://github.com/kubernetes/apiextensions-apiserver/blob/master/pkg/apis/apiextensions/v1beta1/types.go#L193)" version. I don't think this is a must fix in this PR, but let's at least open a tracking issue for handling this better

Reply:
That's interesting.  We may need to store versioned annotations to apply to the appropriate UI and Resource metadata.  Noting this in the PR description as an issue to consider/open.

Reply:
Noted in https://github.com/crossplaneio/crossplane/compare/9f5ab3c6d421d5e1b0d722af3317569614d5132f..017c3d47090daca29f82dfd30b95578a3b3e297d